### DISCUSSION TITLE
How to use NCI-based flake output as input of another?

### DISCUSSION THREAD

**Discussion by nothingnesses** [90-008/nix-cargo-integration#112] [Q&A] [answered]
Hi. I'm trying to use the output of an NCI-based flake as the input of another, but I can't get it to work.

[Here is a repository of a minimal reproduction](https://github.com/nothingnesses/nci-question/tree/34decaea9c090d9f2b5aa178413409bf4fd0e030) where I made a flake for `cargo-generate` which I'm trying to use inside the dev shell of another flake.

Running `nix shell .#cargo-generate` from "./cargo-generate" successfully drops me to a shell where I can use `cargo generate`.

However, running `nix develop` from "./" results in the following error:

```
$ nix develop
error: attribute 'defaultPackage' missing

       at /nix/store/bkgy2dvg91a8578w59534z75nsfy2jbd-source/flake.nix:190:17:

          189|               path = lib.makeBinPath [
          190|                 alejandra.defaultPackage.${system}
             |                 ^
          191|                 pkgs.python3.pkgs.black
(use '--show-trace' to show detailed location information)
$
```

How can I figure out why this is happening and what can I do to get this to work?

**Comment by 90-008** [comment]
It doesn't look like that issue relates to `nix-cargo-integration` but rather to `dream2nix`, as in the code comes from there, but I'm not sure why that would be an error (unless you changed dream2nix's alejandra input?). The repo you linked doesn't seem to be available so I couldn't take a look at it. If you can link a new one here I can take a look at it.

**Reply by nothingnesses** [reply]
Ah thanks. It seems I had the repository set to private which might've been why it was inaccessible to you, so sorry about that. I've just set it to public now, so please do check it again. Please let me know if it's still inaccessible and I'll try to figure out a different way to share it. In any case, I've not fiddled with any `dream2nix`-related settings, afaik.

**Reply by 90-008** [reply]
In this line in `flake.nix`:
```
packages = builtins.map (element: {package = inputs.${element};}) ["cargo-generate"];
```
what you actually want is this I believe
```
packages = builtins.map (element: {package = inputs.${element}.packages.default;}) ["cargo-generate"];
```

The former is mapping to the flake inputs themselves, and not the package.

**Reply by nothingnesses** [reply]
Thanks, but unfortunately that doesn't seem to work:

```
$ nix develop
warning: Git tree '/home/a/tmp/nci' is dirty
error: attribute 'default' missing

       at /nix/store/wlb0fmznr0h3c4r0mj8sjw6vw7z70azn-source/flake.nix:29:56:

           28|         shell = {
           29|           packages = builtins.map (element: {package = inputs.${element}.packages.default;}) ["cargo-generate"];
             |                                                        ^
           30|         };
(use '--show-trace' to show detailed location information)
$
```

**Reply by 90-008** [reply]
You can either set the `outputs.defaults` config option in the `cargo-generate` flake or just replace that line with:
```
packages = [inputs.cargo-generate.packages.cargo-generate];
```

**Reply by nothingnesses** [reply]
Thanks, but I've [already set that config option in the `cargo-generate` flake](https://github.com/nothingnesses/nci-question/blob/34decaea9c090d9f2b5aa178413409bf4fd0e030/cargo-generate/flake.nix#L21).

I tried your other suggestion, but it results in the following error:

```
$ nix develop
warning: Git tree '/home/a/tmp/nci' is dirty
error: attribute 'cargo-generate' missing

       at /nix/store/913a2f27vyh4viyhc6lrgyjwk8kaqh6z-source/flake.nix:29:56:

           28|         shell = {
           29|           packages = builtins.map (element: {package = inputs.${element}.packages.${element};}) ["cargo-generate"];
             |                                                        ^
           30|         };
(use '--show-trace' to show detailed location information)
$ git add .
$ nix develop
warning: Git tree '/home/a/tmp/nci' is dirty
error: attribute 'cargo-generate' missing

       at /nix/store/mg5h7va0ffln9pw6z8iqrw7mxh6sj1yk-source/flake.nix:29:23:

           28|         shell = {
           29|           packages = [inputs.cargo-generate.packages.cargo-generate];
             |                       ^
           30|         };
(use '--show-trace' to show detailed location information)
$
```

**Reply by 90-008** [reply, marked as answer]
Oh sorry, I mixed it up. You need to use `inputs.cargo-generate.packages.${common.pkgs.system}.cargo-generate`. Since this is from the flake inputs, it will be structured like `packages.<system>.<package name>`. I'm actually currently rewriting NCI itself (see #110) to reduce this kind of friction and make it more modular and integrate-able with other stuff. It's pretty close to being merged, so I would be very happy if you could try that, the branch should have a few examples. (though no rendered documentation yet, so you would have to rely on reading the modules code itself, or just DM me on matrix).

**Reply by nothingnesses** [reply]
Thanks. I'll try using that branch. I've been going through the `flake.parts` documentation, but it's quite long, so I probably won't be able to learn enough of it to successfully adapt my example project from above until tomorrow. I'll keep you apprised of my progress.

**Reply by nothingnesses** [reply]
Thanks. I've had to change `inputs.cargo-generate.packages.${common.pkgs.system}.cargo-generate` to `inputs.cargo-generate.packages.${pkgs.system}.default`, as well as fork NCI to add a feature that let me set `--no-default-features` (I've submitted a PR for it) since I kept running into PermissionDenied errors during the build stage of libgit2-sys without that, but other than those, [your solution seems to work](https://github.com/nothingnesses/nci-question/tree/0f81db5206c3a805a76200bea8e4b51b08aa1526).

**Reply by nothingnesses** [reply]
For posterity, the fork is no longer required as [this commit](https://github.com/yusdacra/nix-cargo-integration/commit/9609cd788d82a72087460b28fe45cfa13e4e3d5f) adds a feature that allows `--no-default-features` to be set. [Here's an example usage](https://github.com/nothingnesses/nci-question/tree/b4c6ce4242ce15fb60f7ea8b90bf99dd04f994ee).
